File: 27.c
/* Project Euler, Problem #27
 * Lucas Garron
 * Date: December 26, 2012
 */

#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#include "inc/sieve.h"

#define MAX 999
#define SIEVE_SIZE 1000000

bool* primeQ; // Will be malloced.

bool isPrime(int n) {
  if (n<0) {
    return false;
  }
  assert(n < SIEVE_SIZE);
  return primeQ[n];
}

int primeRun(int a, int b) {
  int n = 0;
  while (isPrime(n*n + a*n+b)) {
    n++;
  }
  return n;
}

int result() {

  primeQ = sieve(SIEVE_SIZE);

  int best = 0;
  int bestCoefficientProduct = 0;

  int a, b;
  for (a = -MAX; a <= MAX; a++) {
    for (b = -MAX; b <= MAX; b++) {
      int run = primeRun(a, b);
      if (run > best) {
        best = run;
        bestCoefficientProduct = a * b;
      }
    }
  }

  free(primeQ);

  return bestCoefficientProduct;
}

int main() {

  printf("%d\n", result());
  
  return 0;
}

File: 27.m
(* Project Euler, Problem #27 *)
(* Lucas Garron *)
(* Date: December 26, 2012 *)

(* Code *)

PrimeRun[a_Integer, b_Integer] := Module[{n = 0},
  While[
    PrimeQ[n^2 + a n + b],
    n++
  ];
  Return[n];
]

max = 0;
maxPair = {};
Do[
  If[
    PrimeRun[a, b] > max,
    max = PrimeRun[a, b]; maxPair = {a, b}
  ],
  {a, -999, 999},
  {b, -999, 999}
]

result = Times @@ maxPair

(* Output *)

Print[result];